File: src/main/java/org/distril/beengine/inventory/impl/CreatureInventory.java
package org.distril.beengine.inventory.impl;

import com.nukkitx.protocol.bedrock.packet.ContainerOpenPacket;
import com.nukkitx.protocol.bedrock.packet.MobArmorEquipmentPacket;
import com.nukkitx.protocol.bedrock.packet.MobEquipmentPacket;
import lombok.Getter;
import org.distril.beengine.entity.EntityCreature;
import org.distril.beengine.inventory.Inventory;
import org.distril.beengine.inventory.InventoryHolder;
import org.distril.beengine.inventory.InventoryType;
import org.distril.beengine.material.item.Item;
import org.distril.beengine.player.Player;
import org.distril.beengine.util.ItemUtils;

import java.util.Collection;

@Getter
public class CreatureInventory extends Inventory {

	private static final int HOTBAR_SIZE = 9;

	private Item helmet, chestplate;
	private Item leggings, boots;

	private int heldItemIndex;

	public CreatureInventory(InventoryHolder holder) {
		super(holder, null);
	}

	public CreatureInventory(InventoryHolder holder, Integer overrideId) {
		super(holder, InventoryType.PLAYER, overrideId);
	}

	@Override
	public void clear() {
		super.clear();

		this.setHelmet(null);
		this.setChestplate(null);
		this.setLeggings(null);
		this.setBoots(null);

		this.setItemInHand(null);
	}

	public Item getItemInHand() {
		return this.getItem(this.getHeldItemIndex());
	}

	public void setItemInHand(Item item) {
		this.setItem(this.getHeldItemIndex(), item);
	}

	public void setHeldItemIndex(int heldItemIndex) {
		if (heldItemIndex >= 0 && heldItemIndex <= HOTBAR_SIZE) {
			this.heldItemIndex = heldItemIndex;
			this.sendHeldItem(this.getHolder().getViewers());
		}
	}

	public void sendHeldItem(Collection<Player> players) {
		var itemInHand = this.getItemInHand();

		var packet = new MobEquipmentPacket();
		packet.setItem(ItemUtils.toNetwork(itemInHand));
		packet.setInventorySlot(this.getHeldItemIndex());
		packet.setHotbarSlot(this.getHeldItemIndex());

		players.forEach(player -> {
			packet.setRuntimeEntityId(this.getHolder().getId());

			if (player.equals(this.getHolder())) {
				packet.setRuntimeEntityId(player.getId());
				this.sendSlot(this.getHeldItemIndex(), player);
			}

			player.sendPacket(packet);
		});
	}

	@Override
	protected void onOpen(Player player) {
		var packet = new ContainerOpenPacket();
		packet.setId((byte) this.getId());
		packet.setType(this.getType().getContainerType());

		var holder = this.getHolder();
		packet.setBlockPosition(holder.getPosition().toInt());
		packet.setUniqueEntityId(holder.getId());

		player.sendPacket(packet);
	}

	public void setHelmet(Item helmet) {
		this.helmet = helmet;
		this.sendArmor();
	}

	public void setChestplate(Item chestplate) {
		this.chestplate = chestplate;
		this.sendArmor();
	}

	public void setLeggings(Item leggings) {
		this.leggings = leggings;
		this.sendArmor();
	}

	public void setBoots(Item boots) {
		this.boots = boots;
		this.sendArmor();
	}

	@Override
	public void sendSlots(Player player) {
		super.sendSlots(player);

		this.sendArmor();
	}

	protected void sendArmor() {
		var holder = this.getHolder();

		var packet = new MobArmorEquipmentPacket();
		packet.setRuntimeEntityId(holder.getId());
		packet.setHelmet(ItemUtils.toNetwork(this.helmet));
		packet.setChestplate(ItemUtils.toNetwork(this.chestplate));
		packet.setLeggings(ItemUtils.toNetwork(this.leggings));
		packet.setBoots(ItemUtils.toNetwork(this.boots));

		holder.getViewers().forEach(viewer -> viewer.sendPacket(packet));
	}

	@Override
	public EntityCreature getHolder() {
		return (EntityCreature) super.getHolder();
	}
}

File: src/main/kotlin/org/distril/beengine/inventory/impl/CreatureInventory.kt
package org.distril.beengine.inventory.impl

import com.nukkitx.protocol.bedrock.packet.ContainerOpenPacket
import com.nukkitx.protocol.bedrock.packet.MobArmorEquipmentPacket
import com.nukkitx.protocol.bedrock.packet.MobEquipmentPacket
import org.distril.beengine.entity.EntityCreature
import org.distril.beengine.inventory.Inventory
import org.distril.beengine.inventory.InventoryHolder
import org.distril.beengine.inventory.InventoryType
import org.distril.beengine.material.item.Item
import org.distril.beengine.player.Player
import org.distril.beengine.util.ItemUtils
import kotlin.reflect.KProperty

open class CreatureInventory(
	holder: InventoryHolder,
	overrideId: Int? = null
) : Inventory(holder, InventoryType.PLAYER, overrideId) {

	override val holder = super.holder as EntityCreature

	var helmet: Item? by ArmorDelegate()
	var chestplate: Item? by ArmorDelegate()
	var leggings: Item? by ArmorDelegate()
	var boots: Item? by ArmorDelegate()

	var heldItemIndex: Int = 0
		set(value) {
			if (value in 0..HOTBAR_SIZE) {
				field = value

				this.sendHeldItem(this.holder.viewers)
			}
		}

	override fun clear() {
		super.clear()

		this.helmet = null
		this.chestplate = null
		this.leggings = null
		this.boots = null

		this.setItemInHand(null)
	}

	fun getItemInHand() = this.getItem(this.heldItemIndex)

	fun setItemInHand(item: Item?) {
		this.setItem(this.heldItemIndex, item)
	}

	fun sendHeldItem(players: Collection<Player>) {
		val packet = MobEquipmentPacket()
		packet.item = ItemUtils.toNetwork(this.getItemInHand())
		packet.inventorySlot = this.heldItemIndex
		packet.hotbarSlot = packet.inventorySlot

		players.forEach {
			packet.runtimeEntityId = this.holder.id
			if (it == this.holder) {
				packet.runtimeEntityId = it.id
				this.sendSlot(this.heldItemIndex, it)
			}

			it.sendPacket(packet)
		}
	}

	override fun onOpen(player: Player) {
		val packet = ContainerOpenPacket()
		packet.id = this.id.toByte()
		packet.type = this.type.containerType

		player.sendPacket(packet)
	}

	override fun sendSlots(player: Player) {
		super.sendSlots(player)

		this.sendArmor()
	}

	protected fun sendArmor() {
		val packet = MobArmorEquipmentPacket()
		packet.runtimeEntityId = this.holder.id
		packet.helmet = ItemUtils.toNetwork(helmet)
		packet.chestplate = ItemUtils.toNetwork(chestplate)
		packet.leggings = ItemUtils.toNetwork(leggings)
		packet.boots = ItemUtils.toNetwork(boots)

		this.holder.viewers.forEach { it.sendPacket(packet) }
	}

	companion object {

		private const val HOTBAR_SIZE = 9
	}

	inner class ArmorDelegate {

		private var field: Item? = null

		operator fun getValue(thisRef: Any?, property: KProperty<*>) = ItemUtils.getAirIfNull(this.field)

		operator fun setValue(thisRef: Any?, property: KProperty<*>, value: Item?) {
			this.field = value

			sendArmor()
		}
	}
}
